### PR TITLE
Add L0 smoothing

### PR DESCRIPTION
**Pull request by zhou-chao**
L0 smoother is a global image operator that sharpens major edges by increasing their steepness while eliminating a manageable degree of low-amplitude structures. It is one of the most influential and useful tools in its kind and has already found many applications, such as edge extraction, clip-art JPEG artifact removal, and non-photorealistic image generation. This method was published in SIGGRAPH Asia 2011 and the corresponding paper [1] is the most cited one among all papers published in the premier SIGGRAPH Asia conference that year.

L0 smoother has been famous and quite popular in both academia and industry image editing fields since it was proposed. It was the first time to introduce to the community such an idea to decompose images with regard to edge properties based on elegant L0 mathematical models. It has been cited by following papers for 250+ times.

For examples and detailed description, please visit http://www.cse.cuhk.edu.hk/leojia/projects/L0smoothing/index.html.

[1] Xu, Li, Cewu Lu, Yi Xu, and Jiaya Jia. "Image smoothing via L 0 gradient minimization." In ACM Transactions on Graphics (TOG), vol. 30, no. 6, p. 174. ACM, 2011.

### CODE REVIEW

File: modules/ximgproc/src/l0_smooth.cpp [hunk 1]
```diff
@@ -0,0 +1,308 @@
...
+		// convert to frequency domain
+		fft(padded, dst);
+	}
+
+	void dftMultiChannel(InputArray src, vector<Mat> &dst){
+		Mat S = src.getMat();
+
+		split(S, dst);
+
+		for(int i = 0; i < S.channels(); i++){
```
Comment (cbalint13) [inline modules/ximgproc/src/l0_smooth.cpp]: * Could be these FFT's done using OpenCV's parallel_for_() & ParallelLoopBody per each channel ?

File: modules/ximgproc/src/l0_smooth.cpp [hunk 2]
```diff
@@ -0,0 +1,308 @@
...
+			fft(dst[i], dst[i]);
+		}
+	}
+
+	void idftMultiChannel(const vector<Mat> &src, OutputArray dst){
+		Mat *channels = new Mat[src.size()];
+
+		for(int i = 0 ; unsigned(i) < src.size(); i++){
+			idft(src[i], channels[i]);
```
Comment (cbalint13) [inline modules/ximgproc/src/l0_smooth.cpp]: Ditto, can it be parallel_for_() & ParallelLoopBody per each channel ?

File: modules/ximgproc/src/l0_smooth.cpp [hunk 3]
```diff
@@ -0,0 +1,308 @@
...
+		aPanels[1] = imaginaryPart;
+
+		Mat rst;
+		merge(aPanels, 2, dst);
+	}
+
+	void divComplexByRealMultiChannel(const vector<Mat> &numer,
+		const vector<Mat> &denom, vector<Mat> &dst)
+	{
+			for(int i = 0; unsigned(i) < numer.size(); i++)
```
Comment (cbalint13) [inline modules/ximgproc/src/l0_smooth.cpp]: * Can be parallel ?

File: modules/ximgproc/src/l0_smooth.cpp [hunk 4]
```diff
@@ -0,0 +1,308 @@
...
+		Mat hGrad, vGrad;
+		filter2D(h, hGrad, -1, Mat(1, 2, CV_32FC1, kernel_inv));
+		filter2D(v, vGrad, -1, Mat(2, 1, CV_32FC1, kernel_inv));
+
+		vector<Mat> hvGradFreq;
+		dftMultiChannel(hGrad+vGrad, hvGradFreq);
+
+		vector<Mat> numer(S.channels());
+		for(int i = 0; i < S.channels(); i++){
+			numer[i] = numerConst[i] + hvGradFreq[i] * beta;
```
Comment (cbalint13) [inline modules/ximgproc/src/l0_smooth.cpp]: Parallel ?

**Comment by cbalint13** [comment]
@zhou-chao , very nice contribution !

**Comment by zhou-chao** [comment]
@cbalint13  Thanks for your comments! I will make it parallel and replace the tabs.

File: modules/ximgproc/src/l0_smooth.cpp [hunk 5]
```diff
@@ -0,0 +1,365 @@
...
+        }
+
+            // power of 2 of the absolute value of the complex
+        Mat pow2absComplex(InputArray src){
+            Mat S = src.getMat();
+
+            Mat sPanels[2];
+            split(S, sPanels);
+
+            return sPanels[0].mul(sPanels[0]) + sPanels[1].mul(sPanels[1]);
```
Comment (vpisarev) [inline modules/ximgproc/src/l0_smooth.cpp]: I believe, magnitude() does the same thing

File: modules/ximgproc/src/l0_smooth.cpp [hunk 6]
```diff
@@ -0,0 +1,365 @@
...
+                Mat hvMag = h.mul(h) + v.mul(v);
+
+                Mat mask;
+                if(S.channels() == 1)
+                {
+                    threshold(hvMag, mask, lambda/beta, 1, THRESH_BINARY);
+                }
+                else if(S.channels() > 1)
+                {
+                    Mat *channels = new Mat[S.channels()];
```
Comment (vpisarev) [inline modules/ximgproc/src/l0_smooth.cpp]: please, use vector<Mat> instead of doing memory management manually

**Comment by vpisarev** [comment]
@zhou-chao, thank you! this is excellent contribution! besides the few minor comments above, it pretty much ready to be merged

**Comment by zhou-chao** [comment]
@vpisarev Thanks for your feedback! I will push a new version soon.

**Comment by zhou-chao** [comment]
If there’s anything we can do to help, please let us know. :D @vpisarev

**Comment by vpisarev** [comment]
@zhou-chao, this is very well made contribution! thank you a lot! :+1:

**Pull request merged**